src/server/pegasus_hotkey_collector.h

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2017, Xiaomi, Inc.  All rights reserved.
+// This source code is licensed under the Apache License Version 2.0, which
+// can be found in the LICENSE file in the root directory of this source tree.
+
+#pragma once
+
+#include <functional>
+#include <dsn/utility/error_code.h>
+#include <rrdb/rrdb_types.h>
+#include <gtest/gtest_prod.h>
+#include <readerwriterqueue/readerwriterqueue.h>
+#include <dsn/dist/fmt_logging.h>
+#include "base/pegasus_utils.h"
+
+namespace pegasus {
+namespace server {
+
+typedef std::shared_ptr<std::unordered_map<int, int>> thread_queue_map;
+
+class hotkey_coarse_data_collector;
+class hotkey_fine_data_collector;
+
+class data_hash_method
+{
+public:
+    data_hash_method()
+    {
+        _data_capture_hash_bucket_num =
+            dsn_config_get_value_uint64("pegasus.server",
+                                        "data_capture_hash_bucket_num",
+                                        37,
+                                        "the number of data capture hash buckets");
+    }
+
+    uint64_t get_bucket_num() { return _data_capture_hash_bucket_num; }
+
+    int hash(const std::string &data)
+    {
+        return std::hash<std::string>{}(data) % _data_capture_hash_bucket_num;
+    };
+
+private:
+    uint64_t _data_capture_hash_bucket_num;
+};
+
+class hotkey_collector
+{
+public:
+    hotkey_collector(dsn::apps::hotkey_type::type hotkey_type, const std::string &replica_name);
+    // after receiving START RPC, start to capture and analyse
+    bool init();
+    // after receiving STOP RPC or timeout, clear historical data
+    void clear();
+    void capture_blob_data(const ::dsn::blob &key);
+    void capture_msg_data(dsn::message_ex **requests_point, const int count);
+    void capture_str_data(const std::string &data);
+    void capture_multi_get_data(const ::dsn::apps::multi_get_request &request,
+                                const ::dsn::apps::multi_get_response &resp);
+    // analyse_data is a periodic task, only valid when _collector_state == COARSE || FINE
+    void analyse_data();
+    std::string get_status();
```

Comment:
This function is not used? If you want to use it, make it as a static function and add `collector_state_set ` parameter would be better. And it's better to define `collector_state_set ` as `enum class` (not `enum`), BTW, remove `_set` postfix.